tests/e2e_tests.c

```diff
@@ -39,61 +49,122 @@ static void	finally(int *out, int *save)
 	close(*save);
 }
 
-static void	file_compare(char *exp_contens, char *act_fname)
+static void	finally_err(int *out, int *save)
+{
+	fflush(stderr);
+    close(*out);
+	dup2(*save, fileno(stderr));
+	close(*out);
+}
+
+static int	file_compare(char *exp_contens, char *act_fname)
 {
 	int fc = open(act_fname, O_RDONLY, 0600);
 	int len_exp = ft_strlen(exp_contens);
 	char *act = ft_calloc(sizeof(char), len_exp * 2 + 1);
 	assert(act);
 	int len_act = read(fc, act, len_exp * 2);
-	assert(len_act == len_exp);
-	int comp_res = ft_strcmp(exp_contens, act);
-	printf("comparison result %i, expected:[%s] (%i symbols)\nactual:[%s] (%i symbols)\n", comp_res, exp_contens, len_exp, act, len_act);
+	assert(len_act - len_exp <= 3); // expected result doesn't store a number
+	int comp_res = ft_strncmp(exp_contens, act, ft_strlen(exp_contens) - 1);
+	#ifdef DEBUG
+	ft_printf("comparison result %i, expected:[%s] (%i symbols)\nactual:[%s] (%i symbols)\n", comp_res, exp_contens, len_exp, act, len_act);
+	#endif
+	assert(comp_res == 0);
+	int i;
+	for (i = len_act - 2; act[i] >= '0' && act[i] <= '9'; i --);
+	char *sub = ft_substr(act, i, len_act);
+	int mallocs = ft_atoi(sub, 0);
+	free(sub);
 	free(act);
 	close(fc);
+	return (mallocs);
 }
 
-int	e2e_tests(void)
+static void	successful_execution(t_testcase *test, int *mallocs)
 {
 	int out, save;
+	system("(rm -r e2e_f && rm e2e.stdout) 2> /dev/null");
+	assert(system("mkdir e2e_f") == 0);
+	assert(system("cp minishell e2e_f/minishell") == 0);
+	catch("e2e.stdout", &out, &save);
+	char *tmp = ft_strjoin("./e2e_f/minishell ", test->cmd);
+	assert(!!tmp);
+	assert(system(tmp) == 0);
+	finally(&out, &save);
+	*mallocs = file_compare(ft_mapss_get(test->exp, "stdout"), "e2e.stdout");
+	t_dlist	*entry;
+	char	*key;
+	entry = test->exp->head;
+	while (entry)
+	{
+		key = ((t_mapss_entry *)entry->content)->key;
+		free(tmp);
+		tmp = ft_strjoin("e2e_f/", key);
+		assert(!!tmp);
+		if (ft_strcmp(key, "stdout") != 0)
+			file_compare(((t_mapss_entry *)entry->content)->value, tmp);
```

Comment:
what case do you handle here in 105 line?

Reply:
I call file compare (compare the expected output from the map with the actual contens of the file), but only if we're talking about files. Stdout is compared separately in line 94. It's done separately because 1) I get a malloc counter from there, 2) it's guaranteed to exist. Line 104 skips stdout so that it's not checked double

Reply:
confused again, a lot of code) then what do you compare in these two different comparisions?

Reply:
Which two?